[Runtime] Fix leak when casting to AnyHashable.

mikeash opened the pull request:
`_dynamicCastToAnyHashable` assumed that `_swift_convertToAnyHashableIndirect` takes its argument at +1, but that is no longer the case.

rdar://problem/44686587

mikeash committed:
[Runtime] Fix leak when casting to AnyHashable.

_dynamicCastToAnyHashable assumed that _swift_convertToAnyHashableIndirect takes its argument at +1, but that is no longer the case.

rdar://problem/44686587

gottesmm commented on test/stdlib/AnyHashableCasts.swift.gyb:
```diff
@@ -118,6 +118,16 @@ AnyHashableCasts.test("${valueExpr} as ${coercedType} as? ${castType}") {
+  do {
+    let tracked = LifetimeTracked(42)
+    let anyHashable = AnyHashable(tracked)
+    let anyObject = anyHashable as AnyObject
```
This is testing two different cases. Aren't there 3 types of cast consumption? I think you will need a full on address only type to test that, no? Maybe my memory is wrong.

rjmccall commented on stdlib/public/runtime/Casting.cpp:
```diff
@@ -635,23 +635,13 @@ static bool _dynamicCastToAnyHashable(OpaqueValue *destination,
-    sourceType->deallocateBufferIn(&buffer);
+  // Destroy the value if requested.
+  if (flags & DynamicCastFlags::TakeOnSuccess) {
+    sourceType->vw_destroy(source);
```
The other way of doing this, by the way, would be to mark the intrinsic parameter as `__owned`.  I don't have an opinion about which one is more important to optimize.

mikeash committed:
[Runtime] Fix leak in tryDynamicCastBoxedSwiftValue.

The call to swift_unboxFromSwiftValueWithType needs to destroy the value on success when requested in the flags.

rdar://problem/44686587

mikeash commented:
@swift-ci please test
